### PR TITLE
#4469 [Manager] Use dispatched actions to switch workspace and pipeline

### PR DESCRIPTION
**Pull request by eechih**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4469 )
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/store/epics/pipeline/switchPipelineEpic.js [hunk 1]
```diff
@@ -14,20 +14,51 @@
...
+export default (action$, state$, { history }) =>
   action$.pipe(
     ofType(actions.switchPipeline.TRIGGER),
-    filter(action => !isEmpty(action.payload)),
-    debounce(() => interval(1000)),
-    switchMap(action => {
-      const { name } = action.payload;
-      return of(actions.switchPipeline.success(name));
+    withLatestFrom(state$),
+    filter(([, state]) => selectors.getWorkspaceName(state)),
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/pipeline/switchPipelineEpic.js:32]: 判斷 ui.workspace.name 是否存在，不存在則不進行 switchPipeline。因為沒有任何一個 workspace 被建立或沒有成功 switchWorkspace 都不應該進行 switchPipeline。

**Comment by eechih** [comment]
> could you briefly explain what is the difference against the previous PR?

判斷 ui.workspace.name 是否存在，不存在則不進行 switchPipeline

File: ohara-manager/client/src/store/epics/pipeline/switchPipelineEpic.js [hunk 2]
```diff
@@ -14,20 +14,51 @@
...
+    switchMap(([action, state]) => {
+      const workspaceName = selectors.getWorkspaceName(state);
+      const pipelineGroup = get(
+        action.payload,
+        'group',
+        hashByGroupAndName(GROUP.WORKSPACE, workspaceName),
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/pipeline/switchPipelineEpic.js:38]: 當 workspaceName 不存在時，call hashByGroupAndName function 會 error。因此先在 switchMap 之前，透過 filter 捨棄 workspaceName 不存在的 action

**Comment by eechih** [comment]
@saivirtue Could you help me to look why the test/manager-api failed? 🙏

**Comment by saivirtue** [comment]
retry manager-api

**Comment by saivirtue** [comment]
retry manager-ut

**Comment by eechih** [comment]
Thanks @saivirtue

**Comment by eechih** [comment]
retry manager-e2e

**Comment by saivirtue** [comment]
> retry manager-e2e

E2E could not run normally since issue #4470 will break the **create workspace** actions for most of the tests...

**Comment by eechih** [comment]
> E2E could not run normally since issue #4470 will break the create workspace actions for most of the tests...

Thanks @saivirtue explain.

@saivirtue @konekoya @wu87988622 This PR is done, please take a review again, thanks.

**Pull request merged**